src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,262 @@
+// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
+#ifdef __linux__
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+#include <goto-programs/goto_model.h>
+
+gdb_apit::gdb_apit(const char *arg)
+  : binary_name(arg), buffer_position(0), last_read_size(0)
+{
+  memset(buffer, 0, MAX_READ_SIZE_GDB_BUFFER);
+}
+
+int gdb_apit::create_gdb_process()
```

Comment:
What are the advantages of spawning a GDB process ourselves vs. connecting to a GDB API?  That would allow us to do things like connect to remote machines, connect to other debuggers that implement the same API and to interface with tools like valgrind.

Reply:
I have done some research before writing this code, but haven't found a general GDB API. I am happy to reconsider this code and use other solutions.
It seems, like you have a particular API in mind. Might you provide some more details?

Reply:
I'm not a GDB expert but:

https://sourceware.org/gdb/current/onlinedocs/gdb/Remote-Debugging.html

makes reference to the "standard GDB remote serial protocol", which I think is the one documented here:

https://sourceware.org/gdb/current/onlinedocs/gdb/Remote-Protocol.html

It seemed like this could be useful for things like this:

https://elinux.org/Debugging_The_Linux_Kernel_Using_Gdb

where you could build calling contexts from a running Linux kernel.

Reply:
Running the GDB process ourselves keeps this a standalone tool on Linux for now. Connecting to a GDB API requires a network debugger running somewhere. So either we have to take care on setting up the debugger in the network or keep a consistent environment for running the memory analyzer.

I think, that might be on the feature request list, but I don't see, that this is the better solution in general.
Apart from that, I don't want to run a program in the debugger. I mainly want to analyze core dump files at the moment, so there is no benefit to do that in the network as I am not running software inside of gdb at the moment.

I see that this might become desirable for SoC systems. My hope is, that, by separating the gdb_api and the analyzer data structure, given another api later that provides the same functionality, you might exchange the debugger in the background using the same bookkeeping.

Reply:
I see your point but I think we're going to need someone who actually has merge rights for this to comment.  @tautschnig ?